### PR TITLE
Remove HELPER_METHOD_FRAMEs from built-in marshalers

### PR DESCRIPTION
**Pull request by jkoritzinsky**
The built-in marshalers that were implemented in native code used FCalls. Change them to use QCalls or (in the case of some simple methods that don't marshal and the `ICustomMarshaler` marshaler), move the implementation to managed code.

### CODE REVIEW

File: src/coreclr/vm/ilmarshalers.cpp [hunk 1]
```diff
@@ -4296,31 +4296,33 @@ void ILNativeArrayMarshaler::EmitNewSavedSizeArgLocal(ILCodeStream* pslILEmit)
     pslILEmit->EmitSTLOC(m_dwSavedSizeArg);
 }
 
-
-FCIMPL4(void, MngdNativeArrayMarshaler::CreateMarshaler, MngdNativeArrayMarshaler* pThis, MethodTable* pMT, UINT32 dwFlags, PCODE pManagedMarshaler)
+extern "C" void QCALLTYPE MngdNativeArrayMarshaler_CreateMarshaler(MngdNativeArrayMarshaler* pThis, MethodTable* pMT, UINT32 dwFlags, PCODE pManagedMarshaler)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ilmarshalers.cpp:4299]: Is there any reason we wouldn't just write this in managed code?

File: src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs [hunk 2]
```diff
@@ -588,89 +588,260 @@ internal struct MarshalerState
...
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern void ConvertContentsToNative(IntPtr pMarshalState, ref object pManagedHome, IntPtr pNativeHome);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "MngdNativeArrayMarshaler_ConvertSpaceToNative")]
+        private static partial void ConvertSpaceToNative(IntPtr pMarshalState, ObjectHandleOnStack pManagedHome, IntPtr pNativeHome);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern void ConvertSpaceToManaged(IntPtr pMarshalState, ref object pManagedHome, IntPtr pNativeHome,
-                                                          int cElements);
+        internal static void ConvertContentsToNative(IntPtr pMarshalState, ref object pManagedHome, IntPtr pNativeHome)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs:605]: All the APIs that take the managed home by `ref` should likey have the `ref` value reset after the QCall. I see you caught the one where it really mattered, but I would recommend do that just to establish the pattern.

File: src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs [hunk 3]
```diff
@@ -588,89 +588,260 @@ internal struct MarshalerState
...
+#pragma warning disable CA1823, IDE0044 // not used by managed code
+            internal IntPtr m_pElementMT;
+            internal IntPtr m_pManagedElementMarshaler;
+            internal IntPtr m_Array;
+            internal Interop.BOOL m_NativeDataValid;
+            internal Interop.BOOL m_BestFitMap;
+            internal Interop.BOOL m_ThrowOnUnmappableChar;
+            internal ushort m_vt;
+            internal uint m_cElements;
+#pragma warning restore CA1823
```
Comment (huoyaoyuan) [inline src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs:672]: Also restore warning for IDE0044?

File: src/coreclr/vm/ilmarshalers.cpp [hunk 4]
```diff
@@ -3901,14 +3899,16 @@ void ILNativeArrayMarshaler::EmitCreateMngdMarshaler(ILCodeStream* pslILEmit)
...
     if (!IsCLRToNative(m_dwMarshalFlags) && IsOut(m_dwMarshalFlags) && IsIn(m_dwMarshalFlags))
     {
-        // Unmanaged->managed in/out is the only case where we expect the native buffer to contain valid data.
-        _ASSERTE((dwFlags & MngdNativeArrayMarshaler::FLAG_NATIVE_DATA_VALID) == 0);
-        dwFlags |= MngdNativeArrayMarshaler::FLAG_NATIVE_DATA_VALID;
+        pslILEmit->EmitLDC(1);
+    }
+    else
+    {
+        pslILEmit->EmitLDC(0);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ilmarshalers.cpp:3910]: Is this `true` / `false`? Can we get these doc'd so we know for sure.

File: src/coreclr/vm/ilmarshalers.cpp [hunk 5]
```diff
@@ -4623,26 +4550,23 @@ FCIMPL3(void, MngdFixedArrayMarshaler::ConvertContentsToNative, MngdFixedArrayMa
             else
             {
                 pMarshaler->ComToOleArray(&arrayRef, pNativeHome, pThis->m_pElementMT, pThis->m_BestFitMap,
-                    pThis->m_ThrowOnUnmappableChar, pThis->m_NativeDataValid, pThis->m_cElements, pThis->m_pManagedElementMarshaler);
+                    pThis->m_ThrowOnUnmappableChar, false, pThis->m_cElements, pThis->m_pManagedElementMarshaler);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ilmarshalers.cpp:4553]:
```suggestion
                    pThis->m_ThrowOnUnmappableChar, FALSE, pThis->m_cElements, pThis->m_pManagedElementMarshaler);
```

**Review by AaronRobinsonMSFT — APPROVED** [review]
LGTM when green.

File: src/coreclr/vm/ilmarshalers.cpp (hunk 4 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/ilmarshalers.cpp:3910]:
```suggestion
        pslILEmit->EmitLDC(1); // true
    }
    else
    {
        pslILEmit->EmitLDC(0); // false
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 6]
```diff
@@ -4296,31 +4294,17 @@ void ILNativeArrayMarshaler::EmitNewSavedSizeArgLocal(ILCodeStream* pslILEmit)
...
-{
-    FCALL_CONTRACT;
+    BEGIN_QCALL;
 
-    HELPER_METHOD_FRAME_BEGIN_0();
+    GCX_COOP();
 
-    BASEARRAYREF arrayRef = (BASEARRAYREF) *pManagedHome;
+    BASEARRAYREF arrayRef = NULL;
+    GCPROTECT_BEGIN(arrayRef);
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:4306]: Nit: You do not need to GCPROTECT the copy of arg in situations like this. It is cheaper to do `pManagedHome.Get` twice instead.

File: src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs [hunk 7]
```diff
@@ -572,105 +572,288 @@ internal static void ClearNative(IntPtr pUnk)
...
+                return;
+            }
+
+            pManagedHome = marshaler.MarshalNativeToManaged(*pNativeHome);
+        }
+
+#pragma warning disable IDE0060 // Remove unused parameter. These APIs need to match a the shape of a "managed" marshaler.
+        internal static void ClearNative(ICustomMarshaler marshaler, ref object pManagedHome, IntPtr* pNativeHome)
+        {
+            // COMPAT: We never pass null to ClearNative.
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs:830]:
```suggestion
            // COMPAT: We never pass null to CleanUpNativeData.
```
?

File: src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs [hunk 8]
```diff
@@ -572,105 +572,288 @@ internal static void ClearNative(IntPtr pUnk)
...
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern void ClearNative(IntPtr pMarshalState, ref object pManagedHome, IntPtr pNativeHome);
+            *pNativeHome = marshaler.MarshalManagedToNative(pManagedHome);
+        }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern void ClearManaged(IntPtr pMarshalState, ref object pManagedHome, IntPtr pNativeHome);
+        internal static void ConvertContentsToManaged(ICustomMarshaler marshaler, ref object pManagedHome, IntPtr* pNativeHome)
+        {
+            // COMPAT: We never pass null to ConvertContentsToManaged.
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs:818]:
```suggestion
            // COMPAT: We never pass null to MarshalNativeToManaged.
```
?

File: src/coreclr/vm/ilmarshalers.cpp [hunk 9]
```diff
@@ -4350,50 +4334,58 @@ FCIMPL3(void, MngdNativeArrayMarshaler::ConvertSpaceToNative, MngdNativeArrayMar
         FillMemory(*pNativeHome, cbArray, 0);
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:4334]: The `GCX_PREEMP` scope can be extended up to here.

**Review by jkotas — APPROVED** [review]
Thank you!

File: src/coreclr/vm/ilmarshalers.cpp [hunk 10]
```diff
@@ -4296,50 +4294,33 @@ void ILNativeArrayMarshaler::EmitNewSavedSizeArgLocal(ILCodeStream* pslILEmit)
...
         if (cbElement == 0)
             COMPlusThrow(kArgumentException, IDS_EE_COM_UNSUPPORTED_SIG);
 
+        GCX_PREEMP();
+
         SIZE_T cbArray;
         if ( (!ClrSafeInt<SIZE_T>::multiply(cElements, cbElement, cbArray)) || cbArray > MAX_SIZE_FOR_INTEROP)
             COMPlusThrow(kArgumentException, IDS_EE_STRUCTARRAYTOOLARGE);
 
         {
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:4323]:
```suggestion
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 11]
```diff
@@ -4296,50 +4294,33 @@ void ILNativeArrayMarshaler::EmitNewSavedSizeArgLocal(ILCodeStream* pslILEmit)
...
-            GCX_PREEMP();
             *pNativeHome = CoTaskMemAlloc(cbArray);
         }
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:4325]:
```suggestion
        *pNativeHome = CoTaskMemAlloc(cbArray);
```

**Comment by jkoritzinsky** [comment]
Failures are known. Merging this in.

**Pull request merged**